### DISCUSSION TITLE
Write 32bit register.

### DISCUSSION THREAD

**Discussion by murik85** [davidrapan/ha-solarman#232] [Q&A]
Is it possible to write a value to a 32-bit register?

**Comment by davidrapan** [comment]
Of course, what you are trying to accomplish?

**Reply by murik85** [reply]
That's weird. Through solarman.write.......register in U32 and S32 registers values are not written (only read).In U16 and S16 there are no problems. Inverter Afore AF6K-SLP.

![read_write](https://github.com/user-attachments/assets/9e80eea5-ab76-41c0-850d-037d520de677)

**Reply by davidrapan** [reply]
You need to write those values separately.

**Reply by murik85** [reply]
I don't get it. Tell me how.

**Reply by davidrapan** [reply]
If you wan to to write 32b value into 2040 you have to write 16b into 2040 and 16b into 2041.

**Reply by murik85** [reply]
Can you give me an example, because I'm trying, but I can't do it.

**Reply by davidrapan** [reply]
Tell me what you are trying to do and I can try. 😉

**Reply by murik85** [reply]
![2040](https://github.com/user-attachments/assets/c6652b60-1853-4a7a-af57-6695454ca3fa)

Record values ​​from 0W to 5000W.

**Reply by davidrapan** [reply]
Show me stored values in each register respectively using built-in action for reading.

**Reply by murik85** [reply]
![20392040](https://github.com/user-attachments/assets/88d0c9d3-cb80-4c04-ab33-2e33ccfaf1e7)

**Reply by davidrapan** [reply]
Because you need to write only values in range 0-5000 you can completely forget about register number 2039 and just write 16b integer into register number 2040. 😉

**Reply by murik85** [reply]
It doesn't record.

![2040](https://github.com/user-attachments/assets/cb646579-4fe5-4c43-b580-2bed3f401e1a)

**Reply by davidrapan** [reply]
Try to use the other action, with modbus function code 6.

**Reply by murik85** [reply]
![111](https://github.com/user-attachments/assets/1c3270a5-11d8-4cd9-af43-962275d8040d)

**Reply by davidrapan** [reply]
What about when you try to set it to lower value? For example 14.

**Reply by murik85** [reply]
Tried it, it didn't work.

**Reply by davidrapan** [reply]
Are you sure that what you have is a proper documentation for your device? What brand and model you have and which yaml profile are you using currently?

Did you check that the value (currently 15) does match value of the configuration item on your device? (To confirm that you are trying to change correct parameters)

**Reply by murik85** [reply]
Yes, the documentation is correct. The register is 100% read-write. The register is written through Node-Red without problems (I switch stick logger to “transparent” mode).

default:
  code: 0x04
  update_interval: 15


parameters:
  - group: Solar
    items:
      - name: "V_pv1"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x022B]
        icon: "mdi:solar-power"

      - name: "V_pv2"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x022E]
        icon: "mdi:solar-power"

      - name: "I_pv1"
        class: "current"
        state_class: "measurement"
        uom: "A"
        scale: 0.01
        rule: 1
        registers: [0x022C]
        icon: "mdi:current-dc"

      - name: "I_pv2"
        class: "current"
        state_class: "measurement"
        uom: "A"
        scale: 0.01
        rule: 1
        registers: [0x022F]
        icon: "mdi:current-dc"

      - name: "P_pv1"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 1
        registers: [0x022D]
        icon: "mdi:solar-power"

      - name: "P_pv2"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 1
        registers: [0x0230]
        icon: "mdi:solar-power"

      - name: "P_pv"
        class: "power"
        state_class: "total"
        uom: "W"
        scale: 1
        rule: 3
        registers: [0x022A]
        icon: "mdi:solar-power"

  - group: Battery
    items:
      - name: "State_bat"
        update_interval: 300
        rule: 1
        registers: [0x07D0]
        isstr: true
        lookup:
          - key: 0
            value: "No Battery"
          - key: 1
            value: "Fault"
          - key: 2
            value: "Sleep"
          - key: 3
            value: "Start"
          - key: 4
            value: "Charging"
          - key: 5
            value: "Discharge"
          - key: 6
            value: "Stop"
          - key: 7
            value: "Wake Up"
        icon: "mdi:battery"

      - name: "P_bat"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 4
        registers: [0x07D8]
        icon: "mdi:battery"

      - name: "V_bat"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x07D4]
        icon: "mdi:battery"

      - name: "I_bat"
        class: "current"
        state_class: "measurement"
        uom: "A"
        scale: 0.01
        rule: 2
        registers: [0x07D5]
        icon: "mdi:current-dc"

      - name: "SOC_bat"
        class: "battery"
        state_class: "measurement"
        uom: "%"
        scale: 1
        rule: 1
        registers: [0x07D2]
        icon: "mdi:battery"

      - name: "T_bat"
        class: "temperature"
        state_class: "measurement"
        uom: "°C"
        scale: 0.1
        rule: 2
        registers: [0x07D1]
        icon: "mdi:battery"

  - group: EPS
    items:
      - name: "V_eps"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x05DF]
        icon: "mdi:transmission-tower"

      - name: "I_eps"
        class: "current"
        state_class: "measurement"
        uom: "A"
        scale: 0.01
        rule: 2
        registers: [0x05E2]
        icon: "mdi:current-ac"

      - name: "P_eps"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 4
        registers: [0x05EA]
        icon: "mdi:home-lightning-bolt"

  - group: Grid
    items:
      - name: "V_bus"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x01F4]
        icon: "mdi:current-dc"

      - name: "I_gfci"
        class: "current"
        state_class: "measurement"
        uom: "mA"
        scale: 1
        rule: 1
        registers: [0x01F7]
        icon: "mdi:current-ac"

      - name: "V_grid"
        class: "voltage"
        state_class: "measurement"
        uom: "V"
        scale: 0.1
        rule: 1
        registers: [0x01FB]
        icon: "mdi:transmission-tower"

      - name: "I_inv"
        class: "current"
        state_class: "measurement"
        uom: "A"
        scale: 0.01
        rule: 2
        registers: [0x01FE]
        icon: "mdi:current-ac"

      - name: "F_grid"
        class: "frequency"
        state_class: "measurement"
        uom: "Hz"
        scale: 0.01
        rule: 1
        registers: [0x0201]
        icon: "mdi:sine-wave"

      - name: "P_inv"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 4
        registers: [0x020B]
        icon: "mdi:home-lightning-bolt"

      - name: "PF_inv"
        class: "power"
        state_class: "measurement"
        uom: ""
        scale: 0.001
        rule: 2
        registers: [0x020E]
        icon: "mdi:home-lightning-bolt"

      - name: "P_grid"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 4
        registers: [0x0212]
        icon: "mdi:home-lightning-bolt"

      - name: "P_load"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 1
        rule: 4
        registers: [0x0224]
        icon: "mdi:home-lightning-bolt"

  - group: Inverter
    items:
      - name: "Running State"
        update_interval: 300
        rule: 1
        registers: [0x09C4]
        isstr: true
        lookup:
          - key: 1
            value: "Stand-by"
          - key: 2
            value: "Start"
          - key: 3
            value: "Run, Grid Connected"
          - key: 4
            value: "Run, Grid Disconnected"
          - key: 5
            value: "Run generator"
          - key: 6
            value: "From Grid"
          - key: 7
            value: "To Grid"
          - key: 8
            value: "Power failure processing"
          - key: 9
            value: "Shutdown"
          - key: 10
            value: "Fault"
          - key: 11
            value: "Upgrading"
          - key: 12
            value: "Aging"
          - key: 13
            value: "Open loop"
          - key: 14
            value: "Sampling calibration"
        icon: "mdi:play-circle-outline"

      - name: "Inverter radiator temperature"
        class: "temperature"
        state_class: "measurement"
        uom: "°C"
        scale: 0.1
        rule: 2
        registers: [0x09D3]
        icon: "mdi:thermometer"

      - name: "Inverter inner temperature"
        class: "temperature"
        state_class: "measurement"
        uom: "°C"
        scale: 0.1
        rule: 2
        registers: [0x09D2]
        icon: "mdi:thermometer"

  - group: Total Energy
    update_interval: 300
    items:
      - name: "E to GRID day"
        class: "energy"
        state_class: "total_increasing"
        uom: "kWh"
        scale: 0.1
        rule: 1
        registers: [0x03EA]
        icon: "mdi:transmission-tower"

      - name: "E to GRID"
        class: "energy"
        state_class: "total_increasing"
        uom: "kWh"
        scale: 0.1
        rule: 3
        registers: [0x03FB]
        icon: "mdi:transmission-tower"

      - name: "E from GRID day"
        class: "energy"
        state_class: "total_increasing"
        uom: "kWh"
        scale: 0.1
        rule: 1
        registers: [0x03EB]
        icon: "mdi:transmission-tower"

      - name: "E from GRID"
        class: "energy"
        state_class: "total_increasing"
        uom: "kWh"
        scale: 0.1
        rule: 3
        registers: [0x03FD]
        icon: "mdi:transmission-tower"

      - name: "E PV day"
        class: "energy"
        state_class: "total"
        uom: "kWh"
        scale: 0.1
        rule: 3
        registers: [0x03EF]
        icon: "mdi:solar-power"

      - name: "E PV"
        class: "energy"
        state_class: "total"
        uom: "kWh"
        scale: 0.1
        rule: 3
        registers: [0x0403]
        icon: "mdi:solar-power"

      - name: "E to Load"
        class: "energy"
        state_class: "total"
        uom: "kWh"
        scale: 0.1
        rule: 3
        registers: [0x03FF]
        icon: "mdi:home-lightning-bolt-outline"

  - group: Alert
    items:
      - name: "Alert BAT"
        class: "enum"
        rule: 3
        registers: [0x0BBD]
        lookup:
          - key: 0x00000000
            value: "No Error"
          - key: 0x00000001
            value: "PcsBatVolOverFault"
          - key: 0x00000002
            value: "PcsBatVolUnderFault"
          - key: 0x00000020
            value: "PcsBatCommunicationFault"
          - key: 0x00000040
            value: "PcsBatTempSensorOpen"

      - name: "Alert GRID"
        class: "enum"
        rule: 3
        registers: [0x0BC1]
        lookup:
          - key: 0x00000000
            value: "No Error"
          - key: 0x00000001
            value: "GridLossFault"
          - key: 0x00000004
            value: "GridInstVoltHighFault"
          - key: 0x00000010
            value: "GridOverVoltFault"
          - key: 0x00000020
            value: "GridUnderVoltFault"
          - key: 0x00000100
            value: "GridOverFreqFault"
          - key: 0x00000200
            value: "GridUnderFreqFault"

      - name: "Alert OFF-GRID"
        class: "enum"
        rule: 3
        registers: [0x0BC3]
        lookup:
          - key: 0x00000000
            value: "No Error"
          - key: 0x00000001
            value: "UpsOVerPowerFault"
          - key: 0x00000002
            value: "GridConflictFault"

      - name: "Alert AC"
        class: "enum"
        rule: 3
        registers: [0x0BC9]
        lookup:
          - key: 0x00000000
            value: "No Error"
          - key: 0x00000001
            value: "HardwareFault"
          - key: 0x00000004
            value: "RInvInstCurrHighFault"
          - key: 0x00000010
            value: "TInvInstCurrHighFault"
          - key: 0x00000080
            value: "LeakCurrFault"
          - key: 0x00000200
            value: "GridRelayFault"
          - key: 0x00000400
            value: "UpsRelayFault"
          - key: 0x00001000
            value: "Relay4Fault"
          - key: 0x00002000
            value: "RUPSInstCurrHighFault"
          - key: 0x00100000
            value: "UPSInvHighVoltFault"
          - key: 0x00200000
            value: "UPSInvLowVoltFault"
          - key: 0x00400000
            value: "UPSInvHighFreqFault"
          - key: 0x00800000
            value: "UPSInvLowFreqFault"

      - name: "Alert DC"
        class: "enum"
        rule: 3
        registers: [0x0BC7]
        lookup:
          - key: 0x00000000
            value: "No Error"
          - key: 0x00010000
            value: "BusAllVoltSwUnderFault"

  - group: Settings
    code: 0x03
    update_interval: 900
    items:
      - name: "BAT Type"
        platform: select
        rule: 1
        registers: [0x0BB8]
        lookup:
          - key: 0
            value: "No Battery"
          - key: 1
            value: "Lithium"
          - key: 2
            value: "Lead Acid"

      - name: "Meter Type"
        platform: select
        rule: 1
        registers: [0x00D3]
        lookup:
          - key: 0
            value: "No Meter"
          - key: 1
            value: "Current Transformer"

**Reply by davidrapan** [reply]
So in the device it's currently set to 15W?

**Reply by murik85** [reply]
Yes

**Reply by davidrapan** [reply]
Your issue is not about 16b or 32b or if the register is writable or not. Stick is simply refusing to accept the value you are trying to set it to and I have really not much idea why... 😕
